### ISSUE TITLE
Dependency Manager in core

### ISSUE THREAD

**Issue by Glavin001** [Unibeautify/unibeautify#58]
> Add support for beautifiers with executables (port from Atom-Beautify)

See https://github.com/Glavin001/atom-beautify/issues/1687 

- [x] Dependencies
  - [x] Node
    - [x] `require` (builtin)
    - [x] `requireg` (global)
    - ~~Project specific (local)~~ (moved to https://github.com/Unibeautify/unibeautify/issues/81 )
  - [x] Executable
    - [x] Configurable program path
    - ~~Docker~~ (moved to https://github.com/Unibeautify/unibeautify/issues/80 )
- [x] Error handling
  - Add the [`errors` for the dependency](https://github.com/Unibeautify/unibeautify/blob/master/src/DependencyManager/Dependency.ts#L6-L25) to https://github.com/Unibeautify/unibeautify/blob/master/src/DependencyManager/Dependency.ts#L32

**Comment by Glavin001** [comment]
I'm thinking of adding a `dependencies` property to the `Beautifier`.

Modelled after VSCode's `launch.json`: https://code.visualstudio.com/docs/editor/debugging#_launch-configurations

```json
{
  "dependencies": [
    {
      "type": "node",
      "name": "Prettier",
      "package": "prettier"
    },
    {
      "type": "exec",
      "name": "PHP-CS-Fixer",
      "program": "php-cs-fixer",
      "version": {
        "parse": (text: string) => {
          try {
            return text.match(/version (.*) by/)[1] + ".0";
          } catch (error) {
            return text.match(/PHP CS Fixer (\d+\.\d+\.\d+)/)[1];
          }
        }
      },
      "docker": {
        "image": "unibeautify/php-cs-fixer"
      }
    },
  ]
}
```

**Comment by Glavin001** [comment]
# Dependency types

## `node`

Node.js package. Can be `require`d from various sources, in order:
1) local (in `node_modules`)
2) global ([`requireg`](https://www.npmjs.com/package/requireg))
3) builtin (such as editor extension)

## `exec`

Execute a non-Node.js module from command-line.
Test installation by checking version.

### `docker`

Fallback to using [Docker](https://www.docker.com/).

**Comment by Glavin001** [comment]
Another idea is:

```typescript
{
  "dependencies": {
    "Prettier": {
      "type": "node",
      "package": "prettier"
    },
    "PHP-CS-Fixer": {
      "type": "exec",
      "program": "php-cs-fixer",
      "version": {
        "parse": (text: string) => {
          try {
            return text.match(/version (.*) by/)[1] + ".0";
          } catch (error) {
            return text.match(/PHP CS Fixer (\d+\.\d+\.\d+)/)[1];
          }
        }
      },
      "docker": {
        "image": "unibeautify/php-cs-fixer"
      }
    }
  }
}
```

**Comment by szeck87** [comment]
I would vote for the first example.

**Comment by Glavin001** [comment]
@szeck87 : Feel free to use the 👍 reaction 😉.

I do as well and am implementing it as such now.

**Comment by Glavin001** [comment]
Pull Request in progress! https://github.com/Unibeautify/unibeautify/pull/63

**Comment by Glavin001** [comment]
WIP code snippet for tests: Not particularly useful yet.

```typescript
/* global suite, test */

//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

// The module 'assert' provides assertion methods from node
const assert = require("assert");

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from "vscode";
import * as extension from "../src";
import { EditProvider } from "../src/EditProvider";
import { Edit } from "../src/diffUtils";
import { LanguageOptionValues } from "unibeautify";
import * as path from "path";

// Defines a Mocha test suite to group tests of similar kind together
suite("EditProvider", function() {

});

function testFormatFile(fileName: string) {
  // Defines a Mocha unit test
  test(`should format file ${fileName}`, function() {
    
  });
}

export function format(file: string, base: vscode.Uri = vscode.workspace.workspaceFolders![0].uri) {
  const absPath = path.join(
      base.fsPath,
      file
  );
  return vscode.workspace.openTextDocument(absPath).then(doc => {
      const text = doc.getText();
      return vscode.window.showTextDocument(doc).then(
          () => {
              console.time(file);
              return vscode.commands
                  .executeCommand('editor.action.formatDocument')
                  .then(() => {
                      console.timeEnd(file);
                      return { result: doc.getText(), source: text };
                  });
          },
          e => console.error(e)
      );
  });
}


function createEditProviderWithConfig(
  config: LanguageOptionValues,
): EditProvider {
  class CustomEditProvider extends EditProvider {
    beautifyOptions(filePath: string): Promise<LanguageOptionValues> {
      return Promise.resolve(config);
    }
  }
  return new CustomEditProvider();
}
```

**Comment by Glavin001** [comment]
Published to v0.3.0!

**Comment by Glavin001** [comment]
Published to v0.12.0
